Command-Line Functionality

lw3266 opened the pull request:
Instead of running the python code, users can now build the tool then input arguments.

Currently there are five arguments with the second argument being optional.

lw3266 committed:
Add files via upload

patch-3; removed parse-args and uses Click for cli; fixed some errors in file path checking.

lw3266 committed:
reorganized files of the most recent commit

lw3266 committed:
changes to Click arguments

lw3266 committed:
Add files via upload

lw3266 committed:
Add files via upload

lw3266 committed:
Delete src/__main__.py

Jul 15th Update
Description: modifying the command line argument parsing using the Click library.

Changes: modified cli.py, comparsion.py, and setup.py. Deleted main.py.

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,55 @@
+
+@click.command()
+@click.argument('input_path', type = click.STRING)
+@click.argument('output_path', nargs = -1, type = click.STRING)
```
You can use `click.option` here

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,55 @@
+@click.command()
+@click.argument('input_path', type = click.STRING)
+@click.argument('output_path', nargs = -1, type = click.STRING)
+@click.argument('clojure_path', type = click.STRING)
```
I'd prefer these to be options too, since it's hard to remember in which order you need to list them. Also, please add a `help` and explain that arguments mean - it's easy to assume that these are paths to input directories or files

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,55 @@
+def validate_paths(input_path: str, output_path: str) -> None:
+    input_path = Path(input_path)
+    
+    if not input_path.is_dir():
```
If the directory exists, but the files we are analyzing are not there, we'll get an error:
`FileNotFoundError: [Errno 2] No such file or directory: 'input\\clojure_data\\atomsClojure.csv'`

So, we should also check if the exist and exit if that is not the case

lw3266 replied:
For clojure data, yes it is reasonable to check at the early stage so that user will not wait until the error shows up.

However, for the other two, the check seems unnecessary. The change I propose is to check whether all three folders at least contains something, will that be helpful?

renatav replied:
My advice would be to test all cases. We don't want a nice error to be raised if one or more of these files is missing. Actually, having only two should be sufficient. However, in that case, we want to skip including the missing one in our comparison

lw3266 replied:
Im sorry that I still didn't fully understand, 

Could you specify having which two would be sufficient? For instance for postIncr comparison, checking for both codeQL and coccinelle would be sufficient as clojure is just one file (all or nothing to compare). The same applies to all other 10 comparisons. ?

Or for each comparison category, either codeQL&clojure or clojure&coccinelle is sufficient? In that case I will use a hashmap or something to keep track of the # of count for each category. If the count is greater than 1 then there will be an output file for that, otherwise no.

Update on the progress: all other feedbacks have been addressed.

renatav replied:
So what I'm thinking is that if you should be able to compare any of them with whichever one you want. So, clojure and CodeQL, clojure and Coccinelle, CodeQL and Coccinelle. So if one set of files is missing, that shouldn't be an error. If there is only one, or none, print a nice error and quit

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,55 @@
+
+    print(f"Input path: {checked_input_path}")
+    print(f"Output path: {checked_output_path}")
+    print(f"CodeQL_path is:{codeql_path}")
```
These shouldn't be necessary, especially if you replace these arguments with options

renatav commented on src/comparsion.py:
```diff
@@ -0,0 +1,143 @@
+    'literal-encoding' : 'literal_encoding'
+}
+
+coccinelle_mapping = {
```
These are the same as `clojure_mapping`, just without some atom names, right? We can just use one map

lw3266 replied:
good catch, the bigger map indeed captures the smaller mapping entirely. Logically that should not cause any issue, would any further verification is needed to make sure that works?

renatav replied:
No further verification needed. I made sure that atoms names in Coccinelle's output are the same as those in Clojure's.

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,55 @@
+@click.argument('input_path', type = click.STRING)
+@click.argument('output_path', nargs = -1, type = click.STRING)
+@click.argument('clojure_path', type = click.STRING)
+@click.argument('codeql_path', type = click.STRING)
```
Doesn't look like we need this `codeql_path`. These inputs do not contain the full paths.

renatav commented on src/comparsion.py:
```diff
@@ -0,0 +1,143 @@
+
+    clojure_master = pd.read_csv(clojure_directory)
+    # clojure_master['file'] = clojure_master['file'].str.split('/').str[-1]
+    for i in range(clojure_master["file"].__len__()):   
```
This technically works, but it's not very pythonic. Avoid using `__len__`, use `len(list)` instead. A much better way here would be to use `regex`: 
```clojure_master['file'] = clojure_master['file'].apply(
        lambda x: re.sub(f'^{re.escape(clojure_system_path)}', '', x, count=1)
    )
```
Tested and it seems to work

renatav commented on src/comparsion.py:
```diff
@@ -0,0 +1,143 @@
+        file_name = f'{key}_comparison.csv'
+        path = Path(output_directory, file_name)
+        df.to_csv(path, index = False)
+        print(f"DataFrame '{key}' has been saved to {file_name}.")
```
Nitpick: No need to say "DataFrame" here.

lw3266 committed:
Update comparsion.py

lw3266 committed:
Update cli.py

lw3266 commented on src/comparsion.py:
```diff
@@ -132,12 +121,16 @@ def compare_and_output(input_directory, output_directory, codeql_system_path, cl
         k["CQL_Column"] = k["CQL_Column"].astype(float)
         dataframes[key] = k.sort_values(by = ["File", "Line"]).reset_index().drop("index", axis = 1)
 
+    # Filter out rows with zero or one "+" for "CLJ", "CQL", and "CNL" combined
```
This part checks whether for an atom there are at least two sources (indicated by '+').

Let me know if this is what we want

lw3266 replied:
Also another thing is that by truncating the relative path, there are still some chunks left. For instance, there might be something.c, and /build/something.c. In our comparison, these two would be treated as different files. However after brief observation there seems to be no such situation. However I could not proof that rigorously.

FNZhao commented on src/cli.py:
```diff
@@ -0,0 +1,46 @@
+    return input_path, output_path
+
+@click.command()
+@click.option('--input', type = str)
```
It might be better to have a 'help' parameter for each option to provide a brief explanation. For example:
`@click.option('--input', type = str, help = ‘Directory containing the files to be compared')`

FNZhao commented on src/cli.py:
```diff
@@ -0,0 +1,46 @@
+# @click.option(help='You will provide five arguments corresponding to 1) input_path, 2) output_path, 3) codeql_path, 4) clojure_path, 5) coccinelle_path. The output_path can be optional by inputting "" indicating an empty string.')
+
+
+def cli(input, output, codeql, clojure, coccinelle):
```
1. CodeQL path is unclear in its results, so it might not be necessary. One simple way I can think of is to just leave a blank after `--codeql`.
2. Commands for comparing the results of two tools are needed (Clojure vs. CodeQL, Clojure vs. Coccinelle, CodeQL vs. Coccinelle).

FNZhao commented on src/cli.py:
```diff
@@ -0,0 +1,46 @@
+
+
+def cli(input, output, codeql, clojure, coccinelle):
+
```
Providing an instruction for this tool is necessary. A docstring could be added here. For example:
```
    """
    Compares and merges data from Clojure and Coccinelle results.
    
    USAGE:

    compare --input [INPUT_DIRECTORY] --clojure [PATH_IN_CLOJURE_RESULT] --coccinelle [PATH_IN_COCCINELLE_RESULT]
    """
```
It will appear when typing the `--help` command.

lw3266 committed:
Update cli.py

 - Added more information to --help.
 - Changed options to accommodate for --whichtwo, that allows for comparison between only two of the three tools.

lw3266 committed:
Update comparsion.py

Two main goals upon request in this commit:
 1. Allows for comparison between two of the three sources (CQL, CNL, and CLJ). It is still possible to include all three like originally.
2. Keep only rows that contain two or more '+', so that we will only see comparisons with more than one tool reporting a positive matching.

How the new features will be explained in the next few comments where I can high light sections of the code.

I found it to be trick to "align" the files and make the detection of two or more + work, as the file paths need to be exactly the same across any two sources.

Thus it is hard to

lw3266 commented on src/comparsion.py:
```diff
@@ -0,0 +1,159 @@
+    'literal-encoding' : 'literal_encoding'
+}
+
+CLJnCQL = ["CLJ", "CQL"]
+CLJnCNL = ["CLJ", "CNL"]
+CQLnCNL = ["CQL", "CNL"]
+all = ["CLJ", "CQL", "CNL"]
```
Simple lists that helps me reindex columns later. Each list contains exactly the names of the two or all options.

lw3266 commented on src/comparsion.py:
```diff
@@ -0,0 +1,159 @@
+    clojure_directory = Path(input_directory, "clojure_data/atomsClojure.csv")
+    coccinelle_directory = Path(input_directory, "coccinelle_data")
+    dataframes = {}
+    which_two = []
+    if whichtwo == "CQLnCNL":
+        which_two = CQLnCNL
+    elif whichtwo == "CLJvCQL":
+        which_two = CLJnCQL
+    elif whichtwo == "CLJvCNL":
+        which_two = CLJnCNL
+    else:
+        which_two = all
```
not much going on, other than assigning the variable which_two to the lists defined earlier

lw3266 commented on src/comparsion.py:
```diff
@@ -0,0 +1,159 @@
+        # These are the index paths of the data we are looking for. Each is located inside nested dictionaries.
+        for i in range(len(json_value["runs"][0]["results"])):
+            s = pd.Series({
+                "Type": filename,
```
I deleted the [:-6] from the original code. I dont understand why but the atom name seems be missing six characters. After this change, the issue is solved but the cause unknown

lw3266 commented on src/comparsion.py:
```diff
@@ -0,0 +1,159 @@
+    # Output merged CSVs to output_directory
+
+    for key, df in dataframes.items():
+        df = df.reindex(columns=["Type", "File", "Line", "CLJ", "CQL", "CNL", "CLJ_Code", "CQL_Column", "CQL_Code", "CNL_Column", "CNL_Code"])
+        df = pd.DataFrame(df)
+        if whichtwo == "CLJvCQL":
+            df = df.drop(['CNL', 'CNL_Column', 'CNL_Code'], axis = 1)
+        elif whichtwo == "CLJvCNL":
+            df = df.drop(['CQL', 'CQL_Column', 'CQL_Code'], axis = 1)
+            None
+        elif whichtwo == "CQJvCNL":
+            df = df.drop(['CLJ', 'CLJ_Code'], axis = 1)
+        else:
+            None
```
Simple mechanism that drops columns not needed depending on the input options.

lw3266 commented on src/comparsion.py:
```diff
@@ -0,0 +1,159 @@
+            df = df.drop(['CLJ', 'CLJ_Code'], axis = 1)
+        else:
+            None
+        
+        df = df[df.apply(count_plus_specific_columns, axis=1, columns=which_two) >= 2]
```
After dropping, we count the how many '+'s are present in the two or three columns (CLJ, CQL, and CNL). Regardless whether the user wants only two or keep all three, in the end we need at least two positives, so the sum >= 2.

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,45 @@
+@click.option('--codeql', type = str, required = True, help = 'Relative path to truncate from file names in the CodeQL data.')
+@click.option('--clojure', type = str, help = 'Relative path to truncate from file names in the Clojure data.')
+@click.option('--coccinelle', type = str, required = True, help = 'Relative path to truncate from file names in the Coccinelle data.')
+@click.option('--whichtwo', type = str, required = True, help = "This option selects two, or all, of the three tools to be compared. Four possible inputs are supported: 'CLJvCQL', 'CLJvCNL', 'CQLnCNL', and 'all'. Anything else will behave as the 'all' option.")
```
I think that it makes sense, as this would allow the user to only compare two inputs, even if all 3 files exist. You can use click.Choice (https://click.palletsprojects.com/en/8.1.x/options/#choice-options). Additionally, we still need to automatically check which input files exist. So, if this is not provided, but only coccinelle and clojure outputs exist, se to CLJvCN.

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,45 @@
+@click.command()
+@click.option('--input', type = str, required = True, help = 'Directory containing three folders of source data.')
+@click.option('--output', type = str, help = 'Directory to store output files. The input is optional.')
+@click.option('--codeql', type = str, required = True, help = 'Relative path to truncate from file names in the CodeQL data.')
```
These names are confusing. How about `--codeql-rel-path`?

lw3266 committed:
Update cli.py

lw3266 committed:
Update comparsion.py

lw3266 commented:
Assuming that the tool will work as intended, one small thing that is obvious from day one is that for the 'Type' column, it feels redundant to keep as we can tell which atom according to the file name. Is it desirable to remove that column to save space?

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+from src.comparsion import compare_and_output
+
+def validate_paths(input_path: str, output_path: str, whichtwo) -> None:
+    input_path = Path(input_path)
```
We should check if this path exists and raise a nice error if that is not the case. No need to check anything else if this path is incorrect

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+
+def validate_paths(input_path: str, output_path: str, whichtwo) -> None:
+    input_path = Path(input_path)
+    codeQL_directory = Path(input_path, "codeQL_data")
```
If I remember correctly, this directory can contain several files, which is why we cannot check if the one file exists like when validating Clojure. We should check if the directory is empty (the easiest way to do this would be to use `glob` and check if there is at least one `*.csv` file

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+    input_path = Path(input_path)
+    codeQL_directory = Path(input_path, "codeQL_data")
+    clojure_directory = Path(input_path, "clojure_data/atomsClojure.csv")
+    coccinelle_directory = Path(input_path, "coccinelle_data")
```
Same here

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+def validate_paths(input_path: str, output_path: str, whichtwo) -> None:
+    input_path = Path(input_path)
+    codeQL_directory = Path(input_path, "codeQL_data")
+    clojure_directory = Path(input_path, "clojure_data/atomsClojure.csv")
```
A future improvement would be to not hard code this name `atomsClojure.csv`, but that does not have to get done in this PR

lw3266 replied:
I assume that for Clojure input, there will still be only one csv file? In that case, it could be an simple fix we just read from that csv file regardless of its name

renatav replied:
That would be sufficient for now. In theory, you could have multiple files in that folder, but since this is for external use, this doesn't have to be perfect

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+    clojure_exists = clojure_directory.exists()
+    coccinelle_exists = coccinelle_directory.is_dir()
+
+    if not codeQL_exists and (whichtwo == "CQLvCNL" or whichtwo == "CLJvCQL" or whichtwo == "all"):
```
It's better to define and use an enum. When using strings like this, it's very easy to introduce a bug due to a small typo. You could also simplify this code by using a list or a dictionary, mapping input type to a boolean specifying if it should exist. This does work when you have only 3 files, but imagine having just one more (and we might develop another tool in the future). It would become difficult to list all combinations. Even with just 3 tools, using a dictionary here make this code more readable

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+    coccinelle_exists = coccinelle_directory.is_dir()
+
+    if not codeQL_exists and (whichtwo == "CQLvCNL" or whichtwo == "CLJvCQL" or whichtwo == "all"):
+        raise FileNotFoundError(f"\033[91mCodeQL source does not exist: \033[0m{codeQL_directory}")
```
Using colors is a nice touch! You could define constants to make this more readable, see https://stackoverflow.com/questions/15580303/python-output-complex-line-with-floats-colored-by-value

lw3266 replied:
It does help the visual to be a lot nicer, but it seems like the way described in Stack makes all the following text that color, prob much more to learn and play with.

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+    if not coccinelle_exists and (whichtwo == "CQLvCNL" or whichtwo == "CLJvCNL" or whichtwo == "all"):
+        raise FileNotFoundError(f"\033[91mCoccinelle source does not exist: \033[0m{coccinelle_directory}")
+    
+    if not whichtwo:
```
Using a dictionary mapping input types to a booleans specifying if the files are expected to exist could simplify this logic too. In fact, you could create this dictionary directly instead of using 3 variables on lines 12 to 14. You could define an enum (CLOJURE, COCCINELLE, CODEQL) and use these values as keys. You could also use the names (strings) directly, but, like I said, I prefer not to hardcode these value as that can lead to bugs. You could then just use `sum(dict.values()) < 2` to check if only one or not even one path exists

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+        else:
+            print(f"\033[91mWARNING: \033[0mOnly one or fewer input source for comparsion, and the output could be meaningless.")
+            sys.exit(0)
+        print(f"whichtwo:", whichtwo)
```
Leftover debug code probably.

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+        elif codeQL_exists and coccinelle_exists:
+            whichtwo = "CQLvCNL"
+        else:
+            print(f"\033[91mWARNING: \033[0mOnly one or fewer input source for comparsion, and the output could be meaningless.")
```
We should handle errors consistently. In some cases, a `FileNotFoundError` is raise, while here, we print and exist early. There is a `try-except` in the command and that is good. Maybe raise an error here too? You can always define a custom error if raising `FileNotFoundError` does not make sense

renatav replied:
Nitpick: if the script's execution has to be terminated, then this is an error and not a warning. That is more important when using loggers

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+            output_path.mkdir(parents = True, exist_ok = True)
+            print(f"\033[34mATTENTION: \033[0mnew folder created at: {output_path}")
+
+    if not output_path:
```
Nitpick: use `else` instead of checking again

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+            print(f"\033[34mATTENTION: \033[0mnew folder created at: {output_path}")
+
+    if not output_path:
+        output_path = Path(Path.cwd(), "output")
```
Nitpick: this `Path.cwd()` is already a `Path` object, so no need to convert to `Path` again. So, `Path.cwd() / "output"` is sufficiet

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+    if not output_path:
+        output_path = Path(Path.cwd(), "output")
+
+        if not output_path.is_dir():        
```
You could define an inner or util function to avoid repeating this code

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+@click.command()
+@click.option('--input', type = str, required = True, help = 'Directory containing three folders of source data.')
+@click.option('--output', type = str, help = 'Directory to store output files. The input is optional.')
+@click.option('--codeql_relative', type = str, required = True, help = 'Relative path to truncate from file names in the CodeQL data.')
```
Nitpick: it's more common to use `-` instead of `_` as a separator in python CLIs. Sorry if I recommended this name

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+@click.option('--coccinelle_relative', type = str, required = True, help = 'Relative path to truncate from file names in the Coccinelle data.')
+@click.option('--whichtwo', type=click.Choice(['CLJvCQL', 'CLJvCNL', 'CQLvCNL', 'all']), help = "This option selects two, or all, of the three tools to be compared. Four possible inputs are supported: 'CLJvCQL', 'CLJvCNL', 'CQLvCNL', and 'all'. If omitted, the program will assign a value to this option by checking input folder availiability.")
+def cli(input, output, codeql_relative, clojure_relative, coccinelle_relative, whichtwo):
+    """The Atoms Comparison tool gathers and compares output from three sources previously developed by the team. Please refer to the README page for more info on each source. To use this project, you will enter four mandatory and two optional options. The usage of each option is listed below. Sample usage: compare --input "/home/usr/Documents/Atoms of Confusion/atoms-comparison-tool" --output "/ ...
```
This needs to be updated since the options were renamed

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+@click.option('--codeql_relative', type = str, required = True, help = 'Relative path to truncate from file names in the CodeQL data.')
+@click.option('--clojure_relative', type = str, help = 'Relative path to truncate from file names in the Clojure data.')
+@click.option('--coccinelle_relative', type = str, required = True, help = 'Relative path to truncate from file names in the Coccinelle data.')
+@click.option('--whichtwo', type=click.Choice(['CLJvCQL', 'CLJvCNL', 'CQLvCNL', 'all']), help = "This option selects two, or all, of the three tools to be compared. Four possible inputs are supported: 'CLJvCQL', 'CLJvCNL', 'CQLvCNL', and 'all'. If omitted, the program will assign a value to this option by checking input folder availiability.")
```
I think that we can simplify this. By default, this should be `None` and we should automatically determine what exists. The most user-friendly way might be to add an option like `exclude`. Then, if you want to exclude CodeQL, you'd run this with `--exclude CodeQL`. Having to type one of these choices is not overly user-friendly. It's not a big deal since this is for our internal use, but would be an issue if that was not the case. And, like I said, imagine one day having 4 tools and needing to support every combination of just 2 and just 3 tools, if not all 4 should be compared. I don't think that this needs to be added in this PR, as long as we can automatically detect that only 2 are present and compared them

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,71 @@
+@click.option('--clojure_relative', type = str, help = 'Relative path to truncate from file names in the Clojure data.')
+@click.option('--coccinelle_relative', type = str, required = True, help = 'Relative path to truncate from file names in the Coccinelle data.')
+@click.option('--whichtwo', type=click.Choice(['CLJvCQL', 'CLJvCNL', 'CQLvCNL', 'all']), help = "This option selects two, or all, of the three tools to be compared. Four possible inputs are supported: 'CLJvCQL', 'CLJvCNL', 'CQLvCNL', and 'all'. If omitted, the program will assign a value to this option by checking input folder availiability.")
+def cli(input, output, codeql_relative, clojure_relative, coccinelle_relative, whichtwo):
```
We should use a more intuitive name. It's really unusual for a command to be called `cli`

renatav commented on src/comparsion.py:
```diff
@@ -0,0 +1,161 @@
+
+    # Load CodeQL .sarif files and store them in dataframes dictionary
+
+    if whichtwo == "CQLvCNL" or whichtwo == "CLJvCQL" or whichtwo == "all":
```
By using a dictionary, you could simplify this check. Something like `if inputs["codeql"]:` would work

renatav commented on src/comparsion.py:
```diff
@@ -0,0 +1,161 @@
+            filename = Path(file_path).stem
+            dataframe_key = filename
+            json_value = json.loads(file_content)
+            p = pd.DataFrame(["Type", "File", "Line", "CQL", "CQL_Column", "CQL_CodeQL"])
```
This code is still in `atom_comparison_tool.py` too, right?  If `atom_comparison_tool.py` is not longer needed, let's remove it

renatav commented on src/comparsion.py:
```diff
@@ -0,0 +1,161 @@
+    'literal-encoding' : 'literal_encoding'
+}
+
+CLJnCQL = ["CLJ", "CQL"]
```
Same here, it would be nice to come up with a better way of handling a missing input. Listing all combinations is always problematic in the long run

renatav commented on src/comparsion.py:
```diff
@@ -0,0 +1,161 @@
+    for key, df in dataframes.items():
+        df = df.reindex(columns=["Type", "File", "Line", "CLJ", "CQL", "CNL", "CLJ_Code", "CQL_Column", "CQL_Code", "CNL_Column", "CNL_Code"])
+        df = pd.DataFrame(df)
+        if whichtwo == "CLJvCQL":
```
If you need an if with elif and no else, it's better to simply write multiple if statements. Additionally, if you ever need to define something with an empty body, you can use `pass`

renatav reviewed (commented):
Sorry for the delay, both @FNZhao and I were preoccupied with other urgent matters. I took a close look at the code, and have some suggestions. I'd like to address some usability and some minor issue before merging it in

lw3266 committed:
Update cli.py

lw3266 committed:
update

lw3266 committed:
Rename input_validation.py to src/input_validation.py

lw3266 committed:
Rename comparsion.py to comparison.py

typo

lw3266 committed:
Update README.md

lw3266 committed:
Create __init__.py

lw3266 committed:
Update README.md

lw3266 committed:
Update cli.py

lw3266 committed:
Update cli.py

renatav commented on README.md:
```diff
@@ -1,11 +1,33 @@
-# atoms_comparison_tool
+logic:
+1. Check input path and read from input path
```
It's a great idea to update the readme! Top-level readme doesn't typically contain a lot of technical details, like there being a dictionary contain certain information. Such very specific explanations can either be moved to docstrings of the appropriate modules, or to something like docs/doc_name.md. This top-level readme should contain general instructions, helping the users install and run this tool

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,103 @@
+    if not input_path.is_dir():
+        raise FileNotFoundError(f"Input path does not exist: {input_path}")
+    
+    def contains_any(string, items):
```
A generic helper function like this one can be moved to a helper or a utils module. Could you also think of more descriptive names of these input parameters?

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,103 @@
+
+input_list = []
+
+def unknown(name):
```
This does not seem necessary?

lw3266 replied:
The input_list stores all the detected and validated input names, in this case I made it global for easy access from other functions. I could try to see if I can simplify it further.

renatav replied:
Is the function needed?

lw3266 replied:
Oh the function, it takes care of name mappings when the key(folder name) is not recognized by the map. I'd prefer to keep it for now for the current mechanism.

lw3266 replied:
Actually, the orginal intention was to return the function from the mapping. But then I changed it to return list of functions. You might be right, I will check that later and remove it if unneeded.

renatav replied:
Just checked, I don't think that it's being used

renatav commented on src/cli.py:
```diff
@@ -0,0 +1,103 @@
+        output_path = Path.cwd() / "output"
+        if not output_path.is_dir():        
+            output_path.mkdir(parents = True, exist_ok = True)
+            print(TEXT['blue'] + "ATTENTION: " + TEXT['reset'] + "Default output folder created at:", output_path)
```
Not in this PR, but just recently I've been working with this library called loguru which makes it easy to elegantly format the outputs. That could be one of the next steps, once this is merged

renatav commented on src/input_validation.py:
```diff
@@ -0,0 +1,19 @@
+        return False
+    return True
+
+def validate_CLJ(path) -> bool:
```
The last two are the same, so you could create a function just name check_if_contains_csv or something like that.

lw3266 replied:
That could work, I am afraid just in case later we need additional logics for the different inputs. But for now I could merge them into one.

renatav replied:
You could still define two functions, but, for now, just call th same helper function

renatav reviewed (commented):
A few more comments based on the code. Have you tested various edge cases? Like empty folders, wrong paths etc.?

lw3266 commented:
> A few more comments based on the code. Have you tested various edge cases? Like empty folders, wrong paths etc.?

Empty folders are for sure, for wrong paths the logic from previous versions of the code should still hold. I have tried the time when output folders need to be created. However if there are any issues please let me know.

renatav commented:
@lw3266 I ran this on my machine and since it's been a while, I forgot exactly what and how I needed to pass in and what the names of the directories should be. Here are a few issues:
- When you run `--help`, while it is quite detailed now, it does not specify what the directories containing codeql, coccinelle and clojure data are expected to be called.
- If these directories are not found (I named mine `CodeQLData`, `ClojureData`), this error is raised: 
```
raise Exception("Only " + str(len(input_list)) + " valid input sources exist.")
Exception: Only 0 valid input sources exist. 
```
  - We can raise the error in our code, but should then catch it and only print the message
  - Ony 0 -> No valid input sources exist or just "Insufficient number of input sources". This error message should also list the names of directories that are expected to be found, so that the user can fix everything just based on that message
  - Use f-strings instead of string concatenation, it's more readable and even slightly faster
- When I fixed that, I started getting this error:
```
(atoms) λ compare --input . --codeql-relative /Users/anuraagpandhi/VSCode/AtomFinder/atom-finder/GitSourceCode --coccinelle-relative /home/ubuntu/atoms/projects
ATTENTION: Input found to include:  ['clojure_data', 'coccinelle_data', 'codeQL_data']
Traceback (most recent call last):
  File "\\?\C:\Users\renat\Envs\atoms\Scripts\compare-script.py", line 33, in <module>
    sys.exit(load_entry_point('AtomsComparsionTool', 'console_scripts', 'compare')())
  File "C:\Users\renat\Envs\atoms\lib\site-packages\click\core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "C:\Users\renat\Envs\atoms\lib\site-packages\click\core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "C:\Users\renat\Envs\atoms\lib\site-packages\click\core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Users\renat\Envs\atoms\lib\site-packages\click\core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "d:\atoms\atoms-comparison-tool\src\cli.py", line 99, in parse_arguments
    func(Path(input, item), relative_mapping.get(item))
  File "d:\atoms\atoms-comparison-tool\src\comparison.py", line 79, in compare_CLJ
    p["file"] = p["file"].apply(lambda x: re.sub(relative, '', x, count=1))
  File "C:\Users\renat\Envs\atoms\lib\site-packages\pandas\core\series.py", line 4924, in apply
    ).apply()
  File "C:\Users\renat\Envs\atoms\lib\site-packages\pandas\core\apply.py", line 1427, in apply
    return self.apply_standard()
  File "C:\Users\renat\Envs\atoms\lib\site-packages\pandas\core\apply.py", line 1507, in apply_standard
    mapped = obj._map_values(
  File "C:\Users\renat\Envs\atoms\lib\site-packages\pandas\core\base.py", line 921, in _map_values
    return algorithms.map_array(arr, mapper, na_action=na_action, convert=convert)
  File "C:\Users\renat\Envs\atoms\lib\site-packages\pandas\core\algorithms.py", line 1743, in map_array
    return lib.map_infer(values, mapper, convert=convert)
  File "lib.pyx", line 2972, in pandas._libs.lib.map_infer
  File "d:\atoms\atoms-comparison-tool\src\comparison.py", line 79, in <lambda>
    p["file"] = p["file"].apply(lambda x: re.sub(relative, '', x, count=1))
  File "C:\Users\renat\AppData\Local\Programs\Python\Python310\lib\re.py", line 209, in sub
    return _compile(pattern, flags).sub(repl, string, count)
  File "C:\Users\renat\AppData\Local\Programs\Python\Python310\lib\re.py", line 302, in _compile
    raise TypeError("first argument must be string or compiled pattern")
TypeError: first argument must be string or compiled pattern
```
Haven't looked into this, but if you started using `pathlib` `Path` instead of strings for example, you'll need to convert that path object back into a string with `str(path)`

lw3266 commented:
> it does not specify what the directories containing codeql, coccinelle and clojure data are expected to be called.

sorry but in that case, we should have standardized the naming in the first place, as I expect that to be the responsibility of the user to be aware of the folder names. 

For now, could I make a standard that all input folder names to be lowercase toolname + "_data"?

In that light, we need to rename the current codeQL folder to codeql.

Let me know your thoughts on that

renatav commented:
@lw3266 You are correct, we will have to improve the naming convention or making it configurable in some way. For now, it would be very helpful if the cli help printed the expected names - or the error message. Something like "Expected to find a directory named codeQL_data". Basically, just let the users know how to name the directory for now. Nothing else needs to be addressed in this PR.

lw3266 commented:
> Haven't looked into this, but if you started using pathlib Path instead of strings for example, you'll need to convert that path object back into a string with str(path)

It was a silly thing, because the --clojure-relative is not entered, the path is a Null object, causing weird behaviors.

This should have been fixed much earlier but I might have accidentally removed the default value while editing other things.

lw3266 committed:
Update input_validation.py

lw3266 committed:
Update cli.py

lw3266 committed:
Update cli.py

lw3266 commented:
Now it should be ready, assuming no other issues occur. 

The addition being the expected_inputs, which takes into account of the available keys (hard coded) with the --except option. For instance, if we have four keys, and the --exclude rules out two of the four, then there will be two items in the expected_output. This does care whether the input folders contains what, just the fullest possible inputs.

1) When the expected_inputs size is less than two, the error will say check your --except, or that your hard coded mapping. Either one of the two is probably very wrong.
2) When the input_list has one item, the error will give the expected_inputs and the input_list, users should make check their input folders to see if any are missing.
3) When the input_list is empty, it makes sense to not print out the list, and just tell that the list is empty, again check the input folders.

lw3266 committed:
Update cli.py

lw3266 committed:
Update cli.py

renatav reviewed (commented):
Looks like there's a syntax error in the newest version:
```
  File "d:\atoms\atoms-comparison-tool\src\cli.py", line 98
    raise InsufficientInputSize(f"{TEXT['red']}Error:{TEXT["reset"]} Check your --exclude input, or the source_mappings in cli.py, the most number of inputs possible is only {len(expected_inputs)}. At least 2 are required.")
                                                            ^^^^^
SyntaxError: f-string: unmatched '['
```
These lines are really long. You usually don't want the lines to contain more than 80-100 characters. You can use a tool like black to format the code. [Take a look](https://black.readthedocs.io/en/stable/)

lw3266 committed:
Update comparison.py

lw3266 commented:
Could you share the input? I did make an update to another file to comparison.py that did cause an error as the names of the functions are not found, which I apologize.

I also tried to update all other files except the README, and Github won't let me commit as they are exactly the same as in my project folder. 

However, I couldn't replicate the syntax error around the square bracket when I raise the InsufficientInputSize error.

renatav commented:
@lw3266  This is what I ran: `compare --input . --codeql-relative /Users/anuraagpandhi/VSCode/AtomFinder/atom-finder/GitSourceCode --coccinelle-relative /home/ubuntu/atoms/projects`

renatav committed:
Formatting, gitgnore, minor fixes and remove jupiter notebook
